### PR TITLE
[2D] Enable 2D optimizer get_state_dict()

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111778
* __->__ #111774

### CODE REVIEW

File: torch/distributed/fsdp/_optim_utils.py [hunk 1]
```diff
@@ -1427,9 +1427,26 @@ def _unflatten_orig_param_states(
...
+                placement_dim = placement.dim  # type: ignore[attr-defined]
+                value_local = value.redistribute(placements=(Replicate(),))
+                reshape_size = list(flat_param._shapes[param_idx])
+                reshape_size[placement_dim] *= 2
+                reshape_size = torch.Size(reshape_size)
+                value = value.reshape(reshape_size)
+            # If gathered state is a replicate DTensor, we directly reshape it.
+            else:
+                value = value.reshape(flat_param._shapes[param_idx])
+        # If gathered state is a tensor, we directly reshape it into unflatten state.
```
Comment (fegin) [inline torch/distributed/fsdp/_optim_utils.py:1446]: This comment should be under `else:` for consistency.

File: torch/distributed/fsdp/_optim_utils.py [hunk 2]
```diff
@@ -1427,9 +1427,26 @@ def _unflatten_orig_param_states(
...
         value = gathered_state[state_name]
-
         param_idx = fsdp_param_info.param_indices[fqn]
-        value = value.reshape(flat_param._shapes[param_idx])
+
+        if isinstance(value, DTensor):
+            placement = value.placements[0]
+            # If gathered state is a DTensor and its TP placement is not Replicate(), we need to
+            # gather the tensor on its TP dimension before chunking them into DTensor again.
+            if placement != Replicate():
```
Comment (fegin) [inline torch/distributed/fsdp/_optim_utils.py:1436]: Let's add a TODO here to mention this solution is not general and is only applicable for PTD TP.

**Review by fegin — APPROVED** [review]
LGTM, please address the comments before landing.

**Comment by wz337** [comment]
> LGTM, please address the comments before landing.

Will address in a following PR.

**Pull request closed without merging**